pkg/query/iter.go

```diff
@@ -304,3 +304,67 @@ func (c *lazySeriesSet) Warnings() annotations.Annotations {
 	}
 	return nil
 }
+
+// histogramResetDetector sets the CounterResetHint to UnknownCounterReset for the first histogram read from the Iterator.
+//
+// The reset hint is always present in the first sample of the chunk, and all
+// consecutive samples will have a NotCounterReset hint. During deduplication the
+// first sample could be skipped, which means the reset will not be properly
+// detected. This iterator will make sure that in that case the hint for the
+// first read sample is set to UnknownCounterReset so that the PromQL engine will
+// do the reset detection manually.
+type histogramResetDetector struct {
+	chunkenc.Iterator
+
+	lastT       int64
+	lastValType chunkenc.ValueType
+
+	i             int16
+	histogramRead bool
+}
+
+func newHistogramResetDetector(iterator chunkenc.Iterator) *histogramResetDetector {
+	return &histogramResetDetector{
+		Iterator: iterator,
+		i:        -1,
```

Comment:
I see the only purpose of this is to check whether Next() has been called? Maybe we could convert it into a simple boolean? Otherwise this might overflow.

Reply:
The variable tracks whether we are at the first sample. We could use a boolean pointer because we need to represent 3 states, one being `Next` not being called at all. AFAIK a chunk has 120 samples so we should not have a risk of overflow, and even if it does the side-effect will be that we recalculate a reset one more time in the engine.